packages/grafana-data/src/types/datasource.ts

```diff
@@ -453,6 +453,8 @@ export interface DataQuery {
    * For non mixed scenarios this is undefined.
    */
   datasource?: string | null;
+
+  syntaxError?: string;
```

Comment:
Not sure putting into this api makes sense as this is to pass data to the data source not the other way around (at least I find it a bit odd to treat this as in/out variable). I think there are 2 options for this:
1. Query editor already know if the query is valid or not. It can directly access the data source and call something like datasource.validate or what ever and just show the error inside the query editor component. Problem is when to do this. Could be on blur but that may be a bit jarring. Ideally we would do it on query run but that isn't controlled in the query editor as the run button is part of explore. I think it could make sense to have some API hook to notify the query editor and do some validation and maybe block the query if it isn't valid. This would be a bit much work as it adds a new API for data sources.
2. Treat this as a query error and just return the response with error from the data source without actually running a request. I think this would be a bit cleaner in the sense of which data are in/out. One downside is that this just treats the error with the same UI as query error. With option 1 you could show the error next to specific input inside the editor. But this is true for the implementation in this PR so I don't think we would loose much.